source/core/slang-relative-container.h

```diff
@@ -0,0 +1,254 @@
+
+    int32_t m_offset;
+};
+
+template <typename T>
+class Safe32Array
+{
+public:
+    const T* begin() const { return m_data; }
+    const T* end() const { return begin() + m_count; }
+
+    T* begin() { return m_data; }
+    T* end() { return begin() + m_count; }
+
+    Index getCount() const { return Index(m_count); }
+    T* getData() { return m_data.get(); }
+    const T* getData() const { return m_data.get(); }
+
+    const T& operator[](Index i) const { SLANG_ASSERT(i >= 0 && uint32_t(i) < m_count); return m_data.get()[i]; }
+    T& operator[](Index i) { SLANG_ASSERT(i >= 0 && uint32_t(i) < m_count); return m_data.get()[i]; }
+
+    Safe32Array(Safe32Ptr<T> data, uint32_t count):m_data(data), m_count(count) {}
+
+    Safe32Array():m_count(0) {}
+
+    Safe32Ptr<T> m_data;
+    uint32_t m_count;
+};
+
+
+template <typename T>
+class Relative32Array
+{
+public:
+    typedef Relative32Array ThisType;
+
+    const T* begin() const { return m_data; }
+    const T* end() const { return begin() + m_count; }
+
+    T* begin() { return m_data; }
+    T* end() { return begin() + m_count; }
+
+    Index getCount() const { return Index(m_count); }
+    T* getData() { return m_data.get(); }
+    const T* getData() const { return m_data.get(); }
+
+    const T& operator[](Index i) const { SLANG_ASSERT(i >= 0 && uint32_t(i) < m_count); return m_data.get()[i]; }
+    T& operator[](Index i) { SLANG_ASSERT(i >= 0 && uint32_t(i) < m_count); return m_data.get()[i]; }
+
+    Relative32Array(const Safe32Array<T>& rhs):
+        m_count(rhs.m_count),
+        m_data(rhs.m_data)
+    {
+    }
+
+    Relative32Array() : m_count(0) {}
+    Relative32Array(const ThisType& rhs) : m_count(rhs.m_count), m_data(rhs.m_data) {}
+
+    uint32_t m_count;               ///< the size of the data
+    Relative32Ptr<T> m_data;           ///< The data
+};
+
+struct RelativeString
+{
+    enum
+    {
+        kSizeBase = 251,
+        kMaxSizeEncodeSize = 5,
+    };
+
+        /// Get contents as a slice
+    UnownedStringSlice getSlice() const;
+        /// Get null terminated string
+    const char* getCstr() const;
+
+        /// Decode the size. Returns the start of the string text, and outSize holds the size (NOT including terminating 0)
+    static const char* decodeSize(const char* in, size_t& outSize);
+
+        /// Returns the amount of bytes used, end encoding in 'encode'
+    static size_t calcEncodedSize(size_t size, uint8_t encode[kMaxSizeEncodeSize]);
+        /// Calculate the total size needed to store the string *including* terminating 0
+    static size_t calcAllocationSize(const UnownedStringSlice& slice);
+
+        /// Calculate the total size needed to store string. Size should be passed *without* terminating 0
+    static size_t calcAllocationSize(size_t size);
+
+    char m_sizeThenContents[1];
+};
+
+class RelativeContainer
+{
+public:
+
+    template <typename T>
+    Safe32Ptr<T> allocate()
```

Comment:
Okay, I get it now (based on my past experience implementing something a bit like this). The `Safe32Ptr` is used during the construction process to both represent the actual pointer to the bytes being filled in (which won't actually be contiguous the way the final binary will be...), and the offset that the data will occupy in the final binary.